File: eel-core/src/main/scala/io/eels/component/hive/HiveSink.scala
package io.eels.component.hive

import com.sksamuel.exts.Logging
import com.typesafe.config.{Config, ConfigFactory}
import io.eels.schema.Schema
import io.eels.{Sink, SinkWriter}
import org.apache.hadoop.fs.FileSystem
import org.apache.hadoop.hive.metastore.IMetaStoreClient

object HiveSink {
  val CaseErrorMsg = "Writing to hive with a schema that contains upper case characters is discouraged because Hive will lowercase all the values. This might lead to subtle case bugs. It is recommended, but not required, that you explicitly convert schemas to lower case before serializing to hive"
}

case class HiveSink(dbName: String,
                    tableName: String,
                    ioThreads: Int = 4,
                    dynamicPartitioning: Option[Boolean] = None,
                    schemaEvolution: Option[Boolean] = None)
                   (implicit fs: FileSystem, client: IMetaStoreClient) extends Sink with Logging {

  val config: Config = ConfigFactory.load()
  val includePartitionsInData = config.getBoolean("eel.hive.includePartitionsInData")
  val bufferSize = config.getInt("eel.hive.bufferSize")
  val schemaEvolutionDefault = config.getBoolean("eel.hive.sink.schemaEvolution")
  val dynamicPartitioningDefault = config.getBoolean("eel.hive.sink.dynamicPartitioning")
  val errorOnUpperCase = config.getBoolean("eel.hive.sink.errorOnUpperCase")

  val ops = new HiveOps(client)

  def withIOThreads(ioThreads: Int): HiveSink = copy(ioThreads = ioThreads)
  def withDynamicPartitioning(partitioning: Boolean): HiveSink = copy(dynamicPartitioning = Some(partitioning))
  def withSchemaEvolution(schemaEvolution: Boolean): HiveSink = copy(schemaEvolution = Some(schemaEvolution))

  private def dialect(): HiveDialect = {
    val format = ops.tableFormat(dbName, tableName)
    logger.debug(s"Table format is $format")
    io.eels.component.hive.HiveDialect(format)
  }

  private def containsUpperCase(schema: Schema): Boolean = schema.fieldNames.forall { it => it == it.toLowerCase() }

  override def writer(schema: Schema): SinkWriter = {
    if (containsUpperCase(schema)) {
      if (errorOnUpperCase)
        sys.error(HiveSink.CaseErrorMsg)
      else
        logger.warn(HiveSink.CaseErrorMsg)
    }

    if (schemaEvolution.contains(true) || schemaEvolutionDefault) {
      // HiveSchemaEvolve(dbName, tableName, schema)
      throw new UnsupportedOperationException("Schema evolution is not yet implemented")
    }

    val metastoreSchema = ops.schema(dbName, tableName)

    new HiveSinkWriter(
      schema,
      metastoreSchema,
      dbName,
      tableName,
      ioThreads,
      dialect(),
      dynamicPartitioning.contains(true) || dynamicPartitioningDefault,
      includePartitionsInData,
      bufferSize
    )
  }
}

File: eel-core/src/main/kotlin/io/eels/component/hive/HiveSink.java
case class HiveSink(private val dbName: String,
                    private val tableName: String,
                    private val ioThreads: Int = 4,
                    private val dynamicPartitioning: Option[Boolean] = None,
                    private val schemaEvolution: Option[Boolean] = None)
                   (implicit fs: FileSystem, client: IMetaStoreClient) extends Sink with StrictLogging {

  val config = ConfigFactory.load()
  val includePartitionsInData = config.getBoolean("eel.hive.includePartitionsInData")
  val bufferSize = config.getInt("eel.hive.bufferSize")
  val SchemaEvolutionDefault = config.getBoolean("eel.hive.sink.schemaEvolution")
  val DynamicPartitioningDefault = config.getBoolean("eel.hive.sink.dynamicPartitioning")

  def withIOThreads(ioThreads: Int): HiveSink = copy(ioThreads = ioThreads)
  def withDynamicPartitioning(partitioning: Boolean): HiveSink = copy(dynamicPartitioning = Some(partitioning))
  def withSchemaEvolution(schemaEvolution: Boolean): HiveSink = copy(schemaEvolution = Some(schemaEvolution))

  private def dialect(implicit client: IMetaStoreClient): HiveDialect = {
    val format = HiveOps.tableFormat(dbName, tableName)
    logger.debug(s"Table format is $format")
    HiveDialect(format)
  }

  private def containsUpperCase(schema: Schema): Boolean = schema.columnNames.forall(col => col == col.toLowerCase)

  override def writer(schema: Schema): SinkWriter = {
    if (containsUpperCase(schema)) {
      logger.warn("Writing to hive with a schema that contains upper case characters, but hive will lower case all field names. This might lead to subtle case bugs. It is recommended, but not required, that you explicitly convert schemas to lower case before serializing to hive")
    }

    if (schemaEvolution.getOrElse(SchemaEvolutionDefault)) {
      HiveSchemaEvolve(dbName, tableName, schema)
    }

    new HiveSinkWriter(
      schema,
      HiveOps.schema(dbName, tableName),
      dbName,
      tableName,
      ioThreads,
      dialect,
      dynamicPartitioning.getOrElse(DynamicPartitioningDefault),
      includePartitionsInData,
      bufferSize
    )
  }
}
